### DISCUSSION TITLE
Noticed a large increase in memory usage with pydantic v2

### DISCUSSION THREAD

**Discussion by accountForIssues** [pydantic/pydantic#6550] [Question]
I use FastAPI (with Pydantic) and uvicorn (with 8 workers) in a docker container with a 2 GB memory limit. Currently in process of migrating/testing the latest versions, I observed that a lot of the workers weren't starting with no errors.

When removing the memory limit, the workers ran fine but the memory usage increased significantly.

The production instance (pydantic 1.10.8 and FastAPI 0.96) uses about 1 GB with 8 workers. Whereas with just 4 workers, the dev instance (with Pydantic v2 and FastAPI 0.100) uses > 2GB.

I ran the "uvicorn run ..." command line with memray to profile the memory usage and this was the summary result.

![image](https://github.com/pydantic/pydantic/assets/52367365/5e3a6519-bcb7-4df8-9aa1-ed6e3d47c3fc)

I have also attached the flamegraph html (as txt) when running with 1 worker. Note that this is on start-up with no incoming requests.

[memray-flamegraph-uvicorn.1.html.txt](https://github.com/pydantic/pydantic/files/12001771/memray-flamegraph-uvicorn.1.html.txt)

Can anyone see what could be the cause ? And whether this is pydantic or FastAPI related ? I am also doing a lot of refactoring so it's possible that it's something that I broke in my code.

**Comment by MattiasOlla** [comment]
I noticed a similiar, if not as dramatic increase in one of our production systems, with a 35% memory increase from 85 MiB to 115 MiB when going from FastAPI 0.99.1 and Pydantic 1.10.10 to FastAPI 0.100.0 and Pydantic 2.0.2, using Python 3.11.4 in a Debian 12 Docker image.

**Comment by samuelcolvin** [comment]
Thanks so much for reporting.

From the screenshot above, it does look like it's related to pydantic, specifically either the validator or serializer used within `TypeAdapter`.

We'll look into it more and come back if we have any questions or intelligent observations.

In the meantime - is there anything in any of your models that might be an obvious candiate - e.g. a very big default value or something?

**Reply by samuelcolvin** [reply]
looks like this is the same as https://github.com/tiangolo/fastapi/discussions/9857.

What we really need is a MRE, although obviously it won't be that "M" to show the difference.

**Comment by samuelcolvin** [comment]
Discuss in on https://github.com/pydantic/pydantic/issues/6620. Closing this to avoid duplication.
